contracts/hooks/SupplyCapAllowedCallerIssuanceHook.sol

```diff
@@ -0,0 +1,159 @@
+        uint256 _supplyCap
+    )
+        public
+    {
+        supplyCap = _supplyCap;
+
+        // Overwrite _owner param of Ownable contract
+        transferOwnership(_initialOwner);
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Adheres to IManagerIssuanceHook interface, and checks to make sure the current issue call won't push total supply over cap.
+     */
+    function invokePreIssueHook(
+        ISetToken _setToken,
+        uint256 _issueQuantity,
+        address _sender,
+        address /*_to*/
+    )
+        external
+        override
+    {
+        _validateAllowedContractCaller(_sender);
+        
+        uint256 totalSupply = _setToken.totalSupply();
+        require(totalSupply.add(_issueQuantity) <= supplyCap, "Supply cap exceeded");
+    }
+
+    /**
+     * Adheres to IManagerIssuanceHook interface
+     */
+    function invokePreRedeemHook(
+        ISetToken _setToken,
+        uint256 _redeemQuantity,
+        address _sender,
+        address _to
+    )
+        external
+        override
+    {}
+
+    /**
+     * ONLY OWNER: Updates supply cap
+     */
+    function updateSupplyCap(uint256 _newCap) external onlyOwner {
+        supplyCap = _newCap;
+        SupplyCapUpdated(_newCap);
+    }
+
+    /**
+     * ONLY OWNER: Toggle ability for passed addresses to call only allowed caller functions
+     *
+     * @param _callers           Array of caller addresses to toggle status
+     * @param _statuses          Array of statuses for each caller
+     */
+    function updateCallerStatus(address[] calldata _callers, bool[] calldata _statuses) external onlyOwner {
+        require(_callers.length == _statuses.length, "Array length mismatch");
+        require(_callers.length > 0, "Array length must be > 0");
+        require(!_callers.hasDuplicate(), "Cannot duplicate callers");
+
+        for (uint256 i = 0; i < _callers.length; i++) {
+            address caller = _callers[i];
+            bool status = _statuses[i];
+            callAllowList[caller] = status;
+            emit CallerStatusUpdated(caller, status);
+        }
+    }
+
+    /**
+     * ONLY OWNER: Toggle whether anyone can call function, bypassing the callAllowlist 
+     *
+     * @param _status           Boolean indicating whether to allow anyone call
+     */
+    function updateAnyoneCallable(bool _status) external onlyOwner {
+        anyoneCallable = _status;
+        emit AnyoneCallableUpdated(_status);
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+     * Validate if passed address is allowed to call function. If anyoneCallable is set to true, anyone can call otherwise needs to be an EOA or 
+     * approved contract address.
+     */
+    function _validateAllowedContractCaller(address _caller) internal view {
+        require(
+            _caller == tx.origin || anyoneCallable || callAllowList[_caller],
+            "Contract not permitted to call"
+        );
```

Comment:
Just a question... in GeneralIndexModule, [the conditions][1] for `anyoneCallable` restrict the contract to EOA only. Are there reasons to be equally restrictive here?

```solidity
    /*
     * Trade must be an EOA if `anyoneTrade` has been enabled for SetToken on the module.
     */
    function _validateOnlyEOAIfUnrestricted(ISetToken _setToken) internal view {
        if(permissionInfo[_setToken].anyoneTrade) {
            require(msg.sender == tx.origin, "Caller must be EOA Address");
        }
    }
```

[1]: https://github.com/SetProtocol/set-protocol-v2/blob/7657f9a9f77f8e639bb6650b1da5496ba2c4080b/contracts/protocol/modules/GeneralIndexModule.sol#L1109-L1116

Reply:
In this case, the intention is for any one of those 3 conditions to be true to be permitted to call. Compared to the GIM, I believe even if anyone is allowed to call, it must be an EOA.

This is how I'm intending the code to work:
anyoneCallable | Is allowed contract | Is EOA | Should revert
-- | -- | -- | --
false | true | false | No
false | false | true | No
true | false | false | No
false | false | false | Yes